lib/puppet/x509/cert_provider.rb

```diff
@@ -146,14 +146,17 @@ def load_private_key(name, required: false)
   # Load a PEM encoded private key.
   #
   # @param pem [String] PEM encoded private key
-  # @return [OpenSSL::PKey::RSA] The private key
-  # @raise [OpenSSL::PKey::RSAError] The `pem` text does not contain a valid key
+  # @return [OpenSSL::PKey::RSA, OpenSSL::PKey::EC] The private key
+  # @raise [OpenSSL::PKey::PKeyError] The `pem` text does not contain a valid key
   # @api private
   def load_private_key_from_pem(pem)
     # set a non-nil passphrase to ensure openssl doesn't prompt
     # but ruby 2.4.0 & 2.4.1 require at least 4 bytes, see
     # https://github.com/ruby/ruby/commit/f012932218fd609f75f9268812df61fb26e2d0f1#diff-40e4270ec386990ac60d7ab5ff8045a4
-    OpenSSL::PKey::RSA.new(pem, '    ')
+    OpenSSL::PKey.read(pem, '    ')
```

Comment:
I assume this comment from the commit message pertains to this change
>Since the saved private key can be either RSA or EC, use the more
>generic `OpenSSL::PKey.read` method which returns the appropriate class
>of private key.


I'm a little confused by it though. How does `OpenSSL::PKey.read` divine the key type from `pem`? This seems like a fairly significant change from generating a new key, as we were doing before

This is also probably a super naive question, so thank you for bearing with me

Reply:
Ah no worries. The PEM file header specifies which kind of key it is. For EC keys, it says https://github.com/puppetlabs/puppet/pull/7437/files#diff-3f2d8fef5ff1e3206db5dc26f1b2026fR14.